src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,170 @@
+                tempArray[index] = list.get(index);
+            }
+            return tempArray;
+        }
+        return null;
+    }
+
+
     @Override
     public void add(T value) {
-
+        int index = cursor;
+        ArrayList.this.add(value, index);
+        cursor = index + 1;
     }
 
+
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int s = size;
+            if (s == array.length) {
+                array = (T[]) grow();
+            }
+            System.arraycopy(array, index,
+                array, index + 1,
+                s - index);
+            array[index] = value;
+            size = s + 1;
+        }
     }
 
+
     @Override
     public void addAll(List<T> list) {
-
+        if (!list.isEmpty()) {
+            T[] tempArray = (T[]) listToArray(list);
+            int newCapacity = size + tempArray.length;
+            array = Arrays.copyOf(array, newCapacity);
+            System.arraycopy(tempArray, 0, array, size, tempArray.length);
+            size = newCapacity;
+        }
     }
 
+
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            return array[index];
+        }
     }
 
+
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            array[index] = value;
+        }
     }
 
+
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException(errorMsg(index));
+        } else {
+            int newSize = size - 1;
+            final T oldValue = get(index);
+            if (newSize > index) {
+                System.arraycopy(array, index + 1, array, index, newSize - index);
+            }
+            size = newSize;
+            array[newSize] = null;
+            return oldValue;
+        }
     }
 
+
     @Override
     public T remove(T element) {
-        return null;
+        int index = 0;
+        found:
+        {
+            if (element == null) {
+                for (; index < size; index++) {
+                    if (array[index] == null) {
+                        remove(index);
+                        break found;
+                    }
+                }
+            } else {
+                for (; index < size; index++) {
+                    if (element.equals(array[index])) {
+                        remove(index);
+                        break found;
+                    }
+                }
+                throw new NoSuchElementException();
+            }
```

Comment:
Removing the `null` element does not make sense. In that logic, `null` will overwrite `null`. Instead, you may add a check at the start of the method: if the given element is `null` => simply return it.

Then, here you have to remove the `if` statement, and its body, remove the `else` declaration, and leave only its body.

Your code may transform in that way:
```suggestion
          if (element == null) {
              return null;
          }
          for (int i = 0; i < size; i++) {
              if (element.equals(array[i])) {
                  return remove(i);
              }
          }
          throw new NoSuchElementException();
```

Also, don't forget to add an informative error message to your exception. See [here](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html#make-your-exceptions-informative).